This patch originates from the Debian project, see https://www.debian.org/

18dbpagesize_binpower.dpatch by  <hesso@pool.math.tu-berlin.de>


Make sure that the pagesize passed to db__set_pagesize() is a power of two.

nvi stores the content of files in BDB database structures. When initiating a
file, it picks a page size for the database to fit the file within 15 pages,
with a minimal page size of 1K and maximal of 10K.

In vanilla nvi, this size is calculated as a multiple of 1024. Modern versions
of BDB, however, require the page size of a database to be a power of two, which
this patch addresses, ridding us of the following message:

  BDB0511 page sizes must be a power-of-2

--- nvi-1.81.6.orig/common/exf.c	2009-03-09 01:48:01.695862889 +0100
+++ nvi-1.81.6/common/exf.c	2009-03-09 10:42:41.147866272 +0100
@@ -249,11 +249,10 @@
 		 * (vi should have good locality) or smaller than 1K.
 		 */
 		psize = ((sb.st_size / 15) + 1023) / 1024;
-		if (psize > 10)
-			psize = 10;
-		if (psize == 0)
-			psize = 1;
-		psize *= 1024;
+		if (psize >= 8) psize=8<<10;
+		else if (psize >= 4) psize=4<<10;
+		else if (psize >= 2) psize=2<<10;
+		else psize=1<<10;
 
 		F_SET(ep, F_DEVSET);
 		ep->mdev = sb.st_dev;
